{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ef2f029e",
   "metadata": {},
   "source": [
    "# Notebook for running stairwayplot2 on eastern snake data\n",
    "\n",
    "This is modified from a notebook for the same purpose from Dr. Arianna Kuhn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "53986969",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# conda environments:\n",
      "#\n",
      "base                     /Users/harrington/opt/miniconda3\n",
      "bioinfo                  /Users/harrington/opt/miniconda3/envs/bioinfo\n",
      "easySFS               *  /Users/harrington/opt/miniconda3/envs/easySFS\n",
      "entropy                  /Users/harrington/opt/miniconda3/envs/entropy\n",
      "fastqc                   /Users/harrington/opt/miniconda3/envs/fastqc\n",
      "gffread                  /Users/harrington/opt/miniconda3/envs/gffread\n",
      "ipyrad                   /Users/harrington/opt/miniconda3/envs/ipyrad\n",
      "macs2                    /Users/harrington/opt/miniconda3/envs/macs2\n",
      "pixy                     /Users/harrington/opt/miniconda3/envs/pixy\n",
      "py-popgen                /Users/harrington/opt/miniconda3/envs/py-popgen\n",
      "qiime2-2022.2            /Users/harrington/opt/miniconda3/envs/qiime2-2022.2\n",
      "splitfasta               /Users/harrington/opt/miniconda3/envs/splitfasta\n",
      "\n",
      "usage: easySFS.py [-h] [-a] -i VCF_NAME -p POPULATIONS [--proj PROJECTIONS]\n",
      "                  [--preview] [-o OUTDIR] [--ploidy PLOIDY] [--prefix PREFIX]\n",
      "                  [--unfolded] [--order POP_ORDER] [--dtype DTYPE]\n",
      "                  [--GQ GQUAL] [-f] [-v]\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help          show this help message and exit\n",
      "  -a                  Keep all snps within each RAD locus (ie. do _not_\n",
      "                      randomly sample 1 snp per locus).\n",
      "  -i VCF_NAME         name of the VCF input file being converted\n",
      "  -p POPULATIONS      Input file containing population assignments per\n",
      "                      individual\n",
      "  --proj PROJECTIONS  List of values for projecting populations down to\n",
      "                      different sample sizes\n",
      "  --preview           Preview the number of segragating sites per population\n",
      "                      for different projection values.\n",
      "  -o OUTDIR           Directory to write output SFS to\n",
      "  --ploidy PLOIDY     Specify ploidy. Default is 2. Only other option is 1 for\n",
      "                      haploid.\n",
      "  --prefix PREFIX     Prefix for all output SFS files names.\n",
      "  --unfolded          Generate unfolded SFS. This assumes that your vcf file\n",
      "                      is accurately polarized.\n",
      "  --order POP_ORDER   Specify the order of populations for the generated sfs.\n",
      "                      Values for --proj should be in this order as well.\n",
      "  --dtype DTYPE       Data type for use in output sfs. Options are `int` and\n",
      "                      `float`. Default is `float`.\n",
      "  --GQ GQUAL          minimum genotype quality tolerated\n",
      "  -f                  Force overwriting directories and existing files.\n",
      "  -v                  Set verbosity. Dump tons of info to the screen\n",
      "\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "# import some stuff\n",
    "import glob\n",
    "import ipyparallel as ipp\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "# For this to work, dadi must be installed\n",
    "# I run this from my easySFS conda environment - has dadi installed - check the current environment\n",
    "!conda env list\n",
    "\n",
    "easySFS = \"/Users/harrington/easySFS/easySFS.py\"\n",
    "# run easySFS and test that it is installed & spits out the options\n",
    "!$easySFS -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ab1dc535",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Acontortrix_p123_v2_25miss_popAcontortrix',\n",
       " 'Acontortrix_p123_v2_25miss_popeast',\n",
       " 'Acontortrix_p123_v2_25miss_popwest',\n",
       " 'Dpunctatus_p123_v3_25missEAST_popDpunctatus',\n",
       " 'Dpunctatus_p123_v3_25missEAST_popcentral',\n",
       " 'Dpunctatus_p123_v3_25missEAST_popnorth',\n",
       " 'Dpunctatus_p123_v3_25missEAST_popsouth',\n",
       " 'Lgetula_p123_v4_25miss_popcent',\n",
       " 'Lgetula_p123_v4_25miss_popeast',\n",
       " 'Lgetula_p123_v4_25miss_popholb',\n",
       " 'Mflagellum_p123_v3_25missEast_popMflagellum',\n",
       " 'Pguttatus_p123_v2_25miss_popemor',\n",
       " 'Pguttatus_p123_v2_25miss_popgut',\n",
       " 'Sdekayi_p123_v4_25miss_popSdekayi',\n",
       " 'abacura_only_popeast',\n",
       " 'abacura_only_popwest',\n",
       " 'erytro_poperytro',\n",
       " 'milks_denovo-92_popelap',\n",
       " 'milks_denovo-92_popgent',\n",
       " 'milks_denovo-92_poptri']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set up some directories\n",
    "analysis_dir = \"/Users/harrington/Active_Research/Ecotone_genomics/GBS_Data/stairwayplot2/analysis/\"\n",
    "pops_dir = \"/Users/harrington/Active_Research/Ecotone_genomics/GBS_Data/Coexp_dryad/stairwayplot2/pop_files/\"\n",
    "vcf_dir = \"/Users/harrington/Active_Research/Ecotone_genomics/GBS_Data/Coexp_dryad/stairwayplot2/filtered_vcf/\"\n",
    "\n",
    "# Populations files for each pop were made with the \"make_pops_SFS_stairway.R\" script\n",
    "all_pops = glob.glob(pops_dir + \"*\")\n",
    "all_pops = sorted([x.split(\"/\")[-1].rsplit(\".\", 1)[0] for x in all_pops])\n",
    "\n",
    "all_pops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "85ad928a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/harrington/Active_Research/Ecotone_genomics/GBS_Data/Coexp_dryad/stairwayplot2/filtered_vcf/Sdekayi_p123_v4_25miss_popSdekayi.vcf\n",
      "/Users/harrington/Active_Research/Ecotone_genomics/GBS_Data/Coexp_dryad/stairwayplot2/pop_files/Sdekayi_p123_v4_25miss_popSdekayi.txt\n",
      "Processing 1 populations - odict_keys(['Sdekayi'])\n",
      "\n",
      "    Running preview mode. We will print out the results for # of segregating sites\n",
      "    for multiple values of projecting down for each population. The dadi\n",
      "    manual recommends maximizing the # of seg sites for projections, but also\n",
      "    a balance must be struck between # of seg sites and sample size.\n",
      "\n",
      "    For each population you should choose the value of the projection that looks\n",
      "    best and then rerun easySFS with the `--proj` flag.\n",
      "    \n",
      "Sdekayi\n",
      "(2, 4317)\t(3, 6423)\t(4, 8135)\t(5, 9606)\t(6, 10965)\t(7, 12201)\t(8, 13383)\t(9, 14463)\t(10, 15530)\t(11, 16532)\t(12, 17518)\t(13, 18456)\t(14, 19379)\t(15, 20259)\t(16, 21131)\t(17, 21974)\t(18, 22805)\t(19, 23608)\t(20, 24402)\t(21, 25165)\t(22, 25928)\t(23, 26672)\t(24, 27408)\t(25, 28116)\t(26, 28828)\t(27, 29515)\t(28, 30204)\t(29, 30862)\t(30, 31530)\t(31, 32165)\t(32, 32815)\t(33, 33419)\t(34, 34052)\t(35, 34596)\t(36, 35212)\t(37, 35580)\t(38, 36179)\t(39, 35986)\t(40, 36565)\t(41, 34211)\t(42, 34749)\t(43, 20747)\t(44, 21061)\t(45, 11567)\t(46, 11735)\t(47, 5476)\t(48, 5553)\t(49, 2056)\t(50, 2083)\t(51, 526)\t(52, 532)\t(53, 80)\t(54, 81)\t\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## prototype the easySFS preview call for one population\n",
    "pop = \"Sdekayi_p123_v4_25miss_popSdekayi\" # define the population\n",
    "in_vcf = vcf_dir + pop + \".vcf\" # define the vcf file location\n",
    "pop_file = pops_dir + pop + \".txt\" # define the population file location\n",
    "print(in_vcf)\n",
    "print(pop_file)\n",
    "\n",
    "!$easySFS -i \"$in_vcf\" -p \"$pop_file\" -a --preview # run easySFS preview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e58dc72d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Acontortrix_p123_v2_25miss_popAcontortrix\n",
      "Acontortrix_p123_v2_25miss_popeast\n",
      "Acontortrix_p123_v2_25miss_popwest\n",
      "Dpunctatus_p123_v3_25missEAST_popDpunctatus\n",
      "Dpunctatus_p123_v3_25missEAST_popcentral\n",
      "Dpunctatus_p123_v3_25missEAST_popnorth\n",
      "Dpunctatus_p123_v3_25missEAST_popsouth\n",
      "Lgetula_p123_v4_25miss_popcent\n",
      "Lgetula_p123_v4_25miss_popeast\n",
      "Lgetula_p123_v4_25miss_popholb\n",
      "Mflagellum_p123_v3_25missEast_popMflagellum\n",
      "Pguttatus_p123_v2_25miss_popemor\n",
      "Pguttatus_p123_v2_25miss_popgut\n",
      "Sdekayi_p123_v4_25miss_popSdekayi\n",
      "abacura_only_popeast\n",
      "abacura_only_popwest\n",
      "erytro_poperytro\n",
      "milks_denovo-92_popelap\n",
      "milks_denovo-92_popgent\n",
      "milks_denovo-92_poptri\n"
     ]
    }
   ],
   "source": [
    "# Get easySFS preview results for all populations\n",
    "preview_dict = {}\n",
    "for pop in all_pops:\n",
    "    print(pop)\n",
    "    in_vcf = vcf_dir + pop + \".vcf\"\n",
    "    pop_file = pops_dir + pop + \".txt\"\n",
    "    preview_dict[pop] = !$easySFS -i \"$in_vcf\" -p \"$pop_file\" -a --preview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3146b84c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Acontortrix_p123_v2_25miss_popAcontortrix \n",
      " (2, 6528)\t(3, 9771)\t(4, 12355)\t(5, 14585)\t(6, 16629)\t(7, 18502)\t(8, 20313)\t(9, 21991)\t(10, 23667)\t(11, 25266)\t(12, 26853)\t(13, 28392)\t(14, 29918)\t(15, 31397)\t(16, 32875)\t(17, 34307)\t(18, 35747)\t(19, 37142)\t(20, 38551)\t(21, 39933)\t(22, 41317)\t(23, 42668)\t(24, 44030)\t(25, 45355)\t(26, 46697)\t(27, 47997)\t(28, 49322)\t(29, 50609)\t(30, 51917)\t(31, 53171)\t(32, 54465)\t(33, 55686)\t(34, 56967)\t(35, 58163)\t(36, 59431)\t(37, 60609)\t(38, 61865)\t(39, 62999)\t(40, 64244)\t(41, 65237)\t(42, 66470)\t(43, 67119)\t(44, 68337)\t(45, 68133)\t(46, 69331)\t(47, 66878)\t(48, 68026)\t(49, 51807)\t(50, 52660)\t(51, 38546)\t(52, 39158)\t(53, 27102)\t(54, 27514)\t(55, 17960)\t(56, 18226)\t(57, 10641)\t(58, 10793)\t(59, 5530)\t(60, 5607)\t(61, 2182)\t(62, 2210)\t(63, 563)\t(64, 570)\t \n",
      "\n",
      "Acontortrix_p123_v2_25miss_popeast \n",
      " (2, 6383)\t(3, 9563)\t(4, 12085)\t(5, 14242)\t(6, 16230)\t(7, 18040)\t(8, 19803)\t(9, 21454)\t(10, 23092)\t(11, 24646)\t(12, 26203)\t(13, 27684)\t(14, 29185)\t(15, 30600)\t(16, 32057)\t(17, 33395)\t(18, 34818)\t(19, 36013)\t(20, 37404)\t(21, 38045)\t(22, 39398)\t(23, 38546)\t(24, 39819)\t(25, 35349)\t(26, 36444)\t(27, 27544)\t(28, 28345)\t(29, 17105)\t(30, 17573)\t(31, 8103)\t(32, 8310)\t(33, 2298)\t(34, 2353)\t \n",
      "\n",
      "Acontortrix_p123_v2_25miss_popwest \n",
      " (2, 5758)\t(3, 8609)\t(4, 10916)\t(5, 12902)\t(6, 14757)\t(7, 16463)\t(8, 18125)\t(9, 19679)\t(10, 21232)\t(11, 22717)\t(12, 24197)\t(13, 25588)\t(14, 27016)\t(15, 28268)\t(16, 29651)\t(17, 30649)\t(18, 31990)\t(19, 31849)\t(20, 33115)\t(21, 30269)\t(22, 31377)\t(23, 23992)\t(24, 24805)\t(25, 14585)\t(26, 15044)\t(27, 6250)\t(28, 6434)\t(29, 1499)\t(30, 1540)\t \n",
      "\n",
      "Dpunctatus_p123_v3_25missEAST_popDpunctatus \n",
      " (2, 11142)\t(3, 16589)\t(4, 20969)\t(5, 24705)\t(6, 28135)\t(7, 31214)\t(8, 34151)\t(9, 36732)\t(10, 39343)\t(11, 41705)\t(12, 44080)\t(13, 46266)\t(14, 48459)\t(15, 50506)\t(16, 52554)\t(17, 54501)\t(18, 56429)\t(19, 58251)\t(20, 60078)\t(21, 61829)\t(22, 63569)\t(23, 65246)\t(24, 66911)\t(25, 68510)\t(26, 70108)\t(27, 71657)\t(28, 73196)\t(29, 74678)\t(30, 76164)\t(31, 77596)\t(32, 79034)\t(33, 80420)\t(34, 81814)\t(35, 83157)\t(36, 84511)\t(37, 85808)\t(38, 87125)\t(39, 88368)\t(40, 89651)\t(41, 90858)\t(42, 92110)\t(43, 93271)\t(44, 94492)\t(45, 95541)\t(46, 96734)\t(47, 97575)\t(48, 98742)\t(49, 99079)\t(50, 100218)\t(51, 99275)\t(52, 100381)\t(53, 95465)\t(54, 96511)\t(55, 74820)\t(56, 75616)\t(57, 57046)\t(58, 57636)\t(59, 42188)\t(60, 42613)\t(61, 30128)\t(62, 30424)\t(63, 20407)\t(64, 20602)\t(65, 12636)\t(66, 12754)\t(67, 7163)\t(68, 7229)\t(69, 3520)\t(70, 3552)\t(71, 1212)\t(72, 1222)\t(73, 173)\t(74, 174)\t \n",
      "\n",
      "Dpunctatus_p123_v3_25missEAST_popcentral \n",
      " (2, 10970)\t(3, 16344)\t(4, 20703)\t(5, 24384)\t(6, 27841)\t(7, 30882)\t(8, 33873)\t(9, 36614)\t(10, 39299)\t(11, 41794)\t(12, 44253)\t(13, 46555)\t(14, 48837)\t(15, 50983)\t(16, 53119)\t(17, 55095)\t(18, 57107)\t(19, 58835)\t(20, 60739)\t(21, 61786)\t(22, 63584)\t(23, 62128)\t(24, 63779)\t(25, 57977)\t(26, 59402)\t(27, 48239)\t(28, 49347)\t(29, 34237)\t(30, 34979)\t(31, 20126)\t(32, 20537)\t(33, 10282)\t(34, 10482)\t(35, 3743)\t(36, 3814)\t(37, 648)\t(38, 660)\t \n",
      "\n",
      "Dpunctatus_p123_v3_25missEAST_popnorth \n",
      " (2, 9227)\t(3, 13762)\t(4, 17169)\t(5, 19920)\t(6, 22363)\t(7, 24481)\t(8, 26458)\t(9, 28236)\t(10, 29926)\t(11, 31461)\t(12, 32953)\t(13, 34289)\t(14, 35634)\t(15, 36723)\t(16, 37953)\t(17, 38430)\t(18, 39557)\t(19, 38656)\t(20, 39674)\t(21, 35747)\t(22, 36611)\t(23, 28229)\t(24, 28869)\t(25, 16621)\t(26, 16976)\t(27, 5281)\t(28, 5387)\t \n",
      "\n",
      "Dpunctatus_p123_v3_25missEAST_popsouth \n",
      " (2, 7250)\t(3, 10167)\t(4, 12660)\t(5, 10710)\t(6, 12071)\t(7, 5043)\t(8, 5492)\t \n",
      "\n",
      "Lgetula_p123_v4_25miss_popcent \n",
      " (2, 14429)\t(3, 21551)\t(4, 27146)\t(5, 31743)\t(6, 36042)\t(7, 39639)\t(8, 43298)\t(9, 45573)\t(10, 48786)\t(11, 47893)\t(12, 50648)\t(13, 43487)\t(14, 45624)\t(15, 32204)\t(16, 33595)\t(17, 17480)\t(18, 18158)\t(19, 5694)\t(20, 5898)\t \n",
      "\n",
      "Lgetula_p123_v4_25miss_popeast \n",
      " (2, 11319)\t(3, 16636)\t(4, 20799)\t(5, 21596)\t(6, 24374)\t(7, 15171)\t(8, 16543)\t \n",
      "\n",
      "Lgetula_p123_v4_25miss_popholb \n",
      " (2, 14194)\t(3, 21190)\t(4, 26689)\t(5, 31257)\t(6, 35480)\t(7, 39239)\t(8, 42845)\t(9, 46134)\t(10, 49357)\t(11, 52372)\t(12, 55328)\t(13, 58134)\t(14, 60891)\t(15, 63504)\t(16, 66104)\t(17, 68599)\t(18, 71072)\t(19, 73431)\t(20, 75799)\t(21, 78024)\t(22, 80302)\t(23, 82383)\t(24, 84583)\t(25, 86306)\t(26, 88435)\t(27, 88941)\t(28, 90990)\t(29, 88477)\t(30, 90400)\t(31, 82196)\t(32, 83894)\t(33, 68523)\t(34, 69878)\t(35, 49943)\t(36, 50890)\t(37, 31110)\t(38, 31676)\t(39, 15644)\t(40, 15917)\t(41, 5117)\t(42, 5202)\t \n",
      "\n",
      "Mflagellum_p123_v3_25missEast_popMflagellum \n",
      " (2, 19660)\t(3, 29401)\t(4, 36806)\t(5, 42673)\t(6, 48225)\t(7, 53082)\t(8, 57796)\t(9, 61895)\t(10, 66107)\t(11, 68877)\t(12, 72728)\t(13, 71088)\t(14, 74566)\t(15, 46449)\t(16, 48421)\t(17, 21200)\t(18, 21997)\t \n",
      "\n",
      "Pguttatus_p123_v2_25miss_popemor \n",
      " (2, 31209)\t(3, 46645)\t(4, 57759)\t(5, 66553)\t(6, 74101)\t(7, 80601)\t(8, 86557)\t(9, 91887)\t(10, 96907)\t(11, 101499)\t(12, 105879)\t(13, 109876)\t(14, 113775)\t(15, 117324)\t(16, 120835)\t(17, 123982)\t(18, 127167)\t(19, 129964)\t(20, 132862)\t(21, 135178)\t(22, 137817)\t(23, 138620)\t(24, 141007)\t(25, 137271)\t(26, 139374)\t(27, 126505)\t(28, 128251)\t(29, 104404)\t(30, 105712)\t(31, 76317)\t(32, 77189)\t(33, 48200)\t(34, 48706)\t(35, 25250)\t(36, 25495)\t(37, 10120)\t(38, 10208)\t(39, 2213)\t(40, 2231)\t \n",
      "\n",
      "Pguttatus_p123_v2_25miss_popgut \n",
      " (2, 21336)\t(3, 31924)\t(4, 39352)\t(5, 45104)\t(6, 49948)\t(7, 54061)\t(8, 57768)\t(9, 61024)\t(10, 64081)\t(11, 66807)\t(12, 69442)\t(13, 71810)\t(14, 74148)\t(15, 76196)\t(16, 78313)\t(17, 80017)\t(18, 81961)\t(19, 83032)\t(20, 84832)\t(21, 84412)\t(22, 86075)\t(23, 82286)\t(24, 83788)\t(25, 74163)\t(26, 75436)\t(27, 57726)\t(28, 58662)\t(29, 33945)\t(30, 34469)\t(31, 11124)\t(32, 11284)\t \n",
      "\n",
      "Sdekayi_p123_v4_25miss_popSdekayi \n",
      " (2, 4317)\t(3, 6423)\t(4, 8135)\t(5, 9606)\t(6, 10965)\t(7, 12201)\t(8, 13383)\t(9, 14463)\t(10, 15530)\t(11, 16532)\t(12, 17518)\t(13, 18456)\t(14, 19379)\t(15, 20259)\t(16, 21131)\t(17, 21974)\t(18, 22805)\t(19, 23608)\t(20, 24402)\t(21, 25165)\t(22, 25928)\t(23, 26672)\t(24, 27408)\t(25, 28116)\t(26, 28828)\t(27, 29515)\t(28, 30204)\t(29, 30862)\t(30, 31530)\t(31, 32165)\t(32, 32815)\t(33, 33419)\t(34, 34052)\t(35, 34596)\t(36, 35212)\t(37, 35580)\t(38, 36179)\t(39, 35986)\t(40, 36565)\t(41, 34211)\t(42, 34749)\t(43, 20747)\t(44, 21061)\t(45, 11567)\t(46, 11735)\t(47, 5476)\t(48, 5553)\t(49, 2056)\t(50, 2083)\t(51, 526)\t(52, 532)\t(53, 80)\t(54, 81)\t \n",
      "\n",
      "abacura_only_popeast \n",
      " (2, 1395)\t(3, 2079)\t(4, 2562)\t(5, 2933)\t(6, 3248)\t(7, 3518)\t(8, 3756)\t(9, 3970)\t(10, 4164)\t(11, 4341)\t(12, 4506)\t(13, 4657)\t(14, 4800)\t(15, 4932)\t(16, 5059)\t(17, 5178)\t(18, 5293)\t(19, 5401)\t(20, 5506)\t(21, 5606)\t(22, 5702)\t(23, 5795)\t(24, 5884)\t(25, 5971)\t(26, 6055)\t(27, 6132)\t(28, 6211)\t(29, 6286)\t(30, 6360)\t(31, 6426)\t(32, 6496)\t(33, 6561)\t(34, 6628)\t(35, 6685)\t(36, 6749)\t(37, 6771)\t(38, 6831)\t(39, 6782)\t(40, 6839)\t(41, 6688)\t(42, 6742)\t(43, 6385)\t(44, 6435)\t(45, 5750)\t(46, 5795)\t(47, 4905)\t(48, 4942)\t(49, 3920)\t(50, 3948)\t(51, 2889)\t(52, 2909)\t(53, 1976)\t(54, 1990)\t(55, 1276)\t(56, 1285)\t(57, 678)\t(58, 682)\t(59, 256)\t(60, 258)\t(61, 83)\t(62, 84)\t \n",
      "\n",
      "abacura_only_popwest \n",
      " (2, 2024)\t(3, 3028)\t(4, 3706)\t(5, 4215)\t(6, 4628)\t(7, 4962)\t(8, 5259)\t(9, 5504)\t(10, 5735)\t(11, 5907)\t(12, 6096)\t(13, 6186)\t(14, 6344)\t(15, 6276)\t(16, 6409)\t(17, 6020)\t(18, 6130)\t(19, 5183)\t(20, 5269)\t(21, 3649)\t(22, 3706)\t(23, 1875)\t(24, 1903)\t(25, 550)\t(26, 558)\t \n",
      "\n",
      "erytro_poperytro \n",
      " (2, 641)\t(3, 959)\t(4, 1180)\t(5, 1346)\t(6, 1490)\t(7, 1608)\t(8, 1716)\t(9, 1812)\t(10, 1900)\t(11, 1980)\t(12, 2054)\t(13, 2121)\t(14, 2186)\t(15, 2238)\t(16, 2295)\t(17, 2340)\t(18, 2390)\t(19, 2419)\t(20, 2464)\t(21, 2481)\t(22, 2522)\t(23, 2458)\t(24, 2494)\t(25, 2232)\t(26, 2263)\t(27, 1295)\t(28, 1311)\t(29, 609)\t(30, 616)\t(31, 250)\t(32, 253)\t(33, 62)\t(34, 63)\t \n",
      "\n",
      "milks_denovo-92_popelap \n",
      " (2, 757)\t(3, 1133)\t(4, 1405)\t(5, 1616)\t(6, 1805)\t(7, 1965)\t(8, 2114)\t(9, 2236)\t(10, 2360)\t(11, 2471)\t(12, 2578)\t(13, 2676)\t(14, 2772)\t(15, 2858)\t(16, 2944)\t(17, 3021)\t(18, 3100)\t(19, 3155)\t(20, 3228)\t(21, 3274)\t(22, 3341)\t(23, 3383)\t(24, 3446)\t(25, 3470)\t(26, 3528)\t(27, 3543)\t(28, 3598)\t(29, 3589)\t(30, 3641)\t(31, 3610)\t(32, 3659)\t(33, 3606)\t(34, 3651)\t(35, 3542)\t(36, 3584)\t(37, 3423)\t(38, 3461)\t(39, 3242)\t(40, 3277)\t(41, 3022)\t(42, 3053)\t(43, 2768)\t(44, 2796)\t(45, 2458)\t(46, 2482)\t(47, 2113)\t(48, 2133)\t(49, 1779)\t(50, 1795)\t(51, 1442)\t(52, 1455)\t(53, 1151)\t(54, 1161)\t(55, 902)\t(56, 909)\t(57, 673)\t(58, 679)\t(59, 493)\t(60, 497)\t(61, 359)\t(62, 362)\t(63, 277)\t(64, 280)\t(65, 207)\t(66, 209)\t(67, 155)\t(68, 156)\t(69, 112)\t(70, 113)\t(71, 79)\t(72, 80)\t(73, 51)\t(74, 52)\t(75, 31)\t(76, 31)\t(77, 23)\t(78, 24)\t(79, 18)\t(80, 18)\t(81, 8)\t(82, 8)\t \n",
      "\n",
      "milks_denovo-92_popgent \n",
      " (2, 920)\t(3, 1372)\t(4, 1702)\t(5, 1968)\t(6, 2195)\t(7, 2391)\t(8, 2570)\t(9, 2726)\t(10, 2875)\t(11, 3012)\t(12, 3140)\t(13, 3254)\t(14, 3368)\t(15, 3467)\t(16, 3568)\t(17, 3653)\t(18, 3746)\t(19, 3821)\t(20, 3905)\t(21, 3961)\t(22, 4039)\t(23, 4093)\t(24, 4165)\t(25, 4187)\t(26, 4254)\t(27, 4249)\t(28, 4312)\t(29, 4249)\t(30, 4307)\t(31, 4213)\t(32, 4267)\t(33, 4095)\t(34, 4145)\t(35, 3934)\t(36, 3979)\t(37, 3716)\t(38, 3756)\t(39, 3351)\t(40, 3386)\t(41, 2924)\t(42, 2954)\t(43, 2540)\t(44, 2565)\t(45, 2144)\t(46, 2165)\t(47, 1792)\t(48, 1808)\t(49, 1429)\t(50, 1441)\t(51, 1092)\t(52, 1102)\t(53, 870)\t(54, 877)\t(55, 676)\t(56, 681)\t(57, 524)\t(58, 528)\t(59, 397)\t(60, 400)\t(61, 316)\t(62, 318)\t(63, 233)\t(64, 235)\t(65, 173)\t(66, 175)\t(67, 129)\t(68, 130)\t(69, 103)\t(70, 104)\t(71, 80)\t(72, 81)\t(73, 60)\t(74, 61)\t(75, 39)\t(76, 39)\t(77, 16)\t(78, 16)\t(79, 8)\t(80, 8)\t \n",
      "\n",
      "milks_denovo-92_poptri \n",
      " (2, 863)\t(3, 1290)\t(4, 1600)\t(5, 1849)\t(6, 2062)\t(7, 2247)\t(8, 2415)\t(9, 2564)\t(10, 2704)\t(11, 2831)\t(12, 2952)\t(13, 3062)\t(14, 3170)\t(15, 3266)\t(16, 3363)\t(17, 3449)\t(18, 3537)\t(19, 3614)\t(20, 3695)\t(21, 3765)\t(22, 3840)\t(23, 3900)\t(24, 3970)\t(25, 4021)\t(26, 4087)\t(27, 4130)\t(28, 4191)\t(29, 4226)\t(30, 4284)\t(31, 4302)\t(32, 4357)\t(33, 4353)\t(34, 4405)\t(35, 4394)\t(36, 4443)\t(37, 4373)\t(38, 4420)\t(39, 4360)\t(40, 4404)\t(41, 4305)\t(42, 4346)\t(43, 4129)\t(44, 4167)\t(45, 3973)\t(46, 4008)\t(47, 3735)\t(48, 3767)\t(49, 3449)\t(50, 3478)\t(51, 3105)\t(52, 3130)\t(53, 2735)\t(54, 2756)\t(55, 2401)\t(56, 2419)\t(57, 2105)\t(58, 2121)\t(59, 1802)\t(60, 1816)\t(61, 1530)\t(62, 1541)\t(63, 1288)\t(64, 1298)\t(65, 1056)\t(66, 1063)\t(67, 857)\t(68, 863)\t(69, 728)\t(70, 733)\t(71, 603)\t(72, 607)\t(73, 501)\t(74, 505)\t(75, 410)\t(76, 413)\t(77, 342)\t(78, 344)\t(79, 297)\t(80, 298)\t(81, 258)\t(82, 259)\t(83, 226)\t(84, 227)\t(85, 188)\t(86, 189)\t(87, 165)\t(88, 166)\t(89, 141)\t(90, 141)\t(91, 117)\t(92, 118)\t(93, 99)\t(94, 100)\t(95, 85)\t(96, 86)\t(97, 76)\t(98, 77)\t(99, 63)\t(100, 63)\t(101, 51)\t(102, 51)\t(103, 44)\t(104, 44)\t(105, 40)\t(106, 40)\t(107, 32)\t(108, 32)\t(109, 27)\t(110, 27)\t(111, 24)\t(112, 24)\t(113, 14)\t(114, 14)\t(115, 9)\t(116, 9)\t(117, 2)\t(118, 2)\t \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Show the easySFS preview results for all pops\n",
    "for k, v in preview_dict.items():\n",
    "    print(k, \"\\n\", v[-3], \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62c5b9f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set projection level for each population. Have to choose the proj level\n",
    "# based on the sampling.....\n",
    "# For stairwayplot 14 or more samples give nice looking output sfs plots - but don't always have this level of sampling\n",
    "\n",
    "proj_dict = {}\n",
    "proj_dict[\"Acontortrix_p123_v2_25miss_popAcontortrix\"] = 46 # all A. contortrix combined\n",
    "\n",
    "proj_dict[\"Acontortrix_p123_v2_25miss_popeast\"] = 24\n",
    "proj_dict[\"Acontortrix_p123_v2_25miss_popwest\"] = 22\n",
    "\n",
    "proj_dict[\"Dpunctatus_p123_v3_25missEAST_popDpunctatus\"] = 52 # all D. punctatus combined\n",
    "\n",
    "proj_dict[\"Dpunctatus_p123_v3_25missEAST_popcentral\"] = 28\n",
    "proj_dict[\"Dpunctatus_p123_v3_25missEAST_popnorth\"] = 22\n",
    "proj_dict[\"Dpunctatus_p123_v3_25missEAST_popsouth\"] = 6\n",
    "\n",
    "proj_dict[\"Lgetula_p123_v4_25miss_popeast\"] = 6\n",
    "proj_dict[\"Lgetula_p123_v4_25miss_popholb\"] = 30\n",
    "proj_dict[\"Lgetula_p123_v4_25miss_popcent\"] = 14\n",
    "proj_dict[\"Mflagellum_p123_v3_25missEast_popMflagellum\"] = 14\n",
    "proj_dict[\"Pguttatus_p123_v2_25miss_popemor\"] = 30\n",
    "proj_dict[\"Pguttatus_p123_v2_25miss_popgut\"] = 24\n",
    "proj_dict['Sdekayi_p123_v4_25miss_popSdekayi'] = 40\n",
    "proj_dict[\"abacura_only_popeast\"] = 44\n",
    "proj_dict[\"abacura_only_popwest\"] = 20\n",
    "proj_dict[\"erytro_poperytro\"] = 23\n",
    "proj_dict[\"milks_denovo-92_popelap\"] = 42\n",
    "proj_dict[\"milks_denovo-92_poptri\"] = 50\n",
    "proj_dict[\"milks_denovo-92_popgent\"] = 42\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bce58d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## prototype the easySFS proj call for one population\n",
    "pop = \"Sdekayi_p123_v4_25miss_popSdekayi\"\n",
    "pop_analysis_dir = analysis_dir + pop + \"/sfs\"\n",
    "in_vcf = vcf_dir + \"{pop}.vcf\".format(pop=pop)\n",
    "pop_file = pops_dir + \"{}.txt\".format(pop)\n",
    "if not os.path.exists(pop_analysis_dir):\n",
    "    os.makedirs(pop_analysis_dir)\n",
    "    \n",
    "!$easySFS -i \"$in_vcf\" -p \"$pop_file\" -a --proj 46 -o \"$pop_analysis_dir\" -f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12669087",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get easySFS proj results for all populations\n",
    "preview_dict = {}\n",
    "for pop in all_pops:\n",
    "    print(pop)\n",
    "    in_vcf = vcf_dir + \"{pop}.vcf\".format(pop=pop)\n",
    "    pop_file = pops_dir + \"{}.txt\".format(pop)\n",
    "    pop_analysis_dir = analysis_dir + pop + \"/sfs\"\n",
    "    if not os.path.exists(pop_analysis_dir):\n",
    "        os.makedirs(pop_analysis_dir)\n",
    "    proj = proj_dict[pop]\n",
    "    !$easySFS -i \"$in_vcf\" -p \"$pop_file\" -a --proj $proj -o \"$pop_analysis_dir\" -f"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12922a28",
   "metadata": {},
   "source": [
    "### make blueprint files for each population\n",
    "\n",
    "here, using same rates as I did in the C. ruber paper, plus generation time of 3 years:\n",
    "\n",
    "2.2*10-9 / year\n",
    "\n",
    "assume generation time of 3 years\n",
    "\n",
    "->> 6.6e-9 per generation\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3230eb27",
   "metadata": {},
   "outputs": [],
   "source": [
    "blueprint_template = \"\"\"#example blueprint file\n",
    "#input setting\n",
    "popid: {pop} # id of the population (no white space)\n",
    "nseq: {nseqs} # number of sequences\n",
    "L: {seq_length} # total number of observed nucleic sites, including polymorphic and monomorphic\n",
    "whether_folded: true # whethr the SFS is folded (true or false)\n",
    "SFS: {sfs} # snp frequency spectrum: number of singleton, number of doubleton, etc. (separated by white space)\n",
    "#smallest_size_of_SFS_bin_used_for_estimation: 1 # default is 1; to ignore singletons, uncomment this line and change this number to 2\n",
    "#largest_size_of_SFS_bin_used_for_estimation: 15 # default is nseq/2 for folded SFS\n",
    "pct_training: 0.67 # percentage of sites for training\n",
    "nrand: 7 15 22 28 # number of random break points for each try (separated by white space)\n",
    "project_dir: '/Users/harrington/Active_Research/Ecotone_genomics/GBS_Data/stairwayplot2/analysis/{pop}/{nseqs}_samples' # project directory\n",
    "stairway_plot_dir: /Applications/stairway_plot_v2.1.1/stairway_plot_es # directory to the stairway plot files\n",
    "ninput: 200 # number of input files to be created for each estimation\n",
    "#random_seed: 666\n",
    "#output setting\n",
    "mu: 6.6e-9 # assumed mutation rate per site per generation\n",
    "year_per_generation: 3 # assumed generation time (in years)\n",
    "#plot setting\n",
    "plot_title: {pop} # title of the plot\n",
    "xrange: 0.1,10000 # Time (1k year) range; format: xmin,xmax; \"0,0\" for default\n",
    "yrange: 0,0 # Ne (1k individual) range; format: xmin,xmax; \"0,0\" for default\n",
    "xspacing: 2 # X axis spacing\n",
    "yspacing: 2 # Y axis spacing\n",
    "fontsize: 12 # Font size\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cd0be91",
   "metadata": {},
   "outputs": [],
   "source": [
    "stairwayplot_es_dir = \"/Applications/stairway_plot_v2.1.1/stairway_plot_es\"\n",
    "\n",
    "# prototype for one population\n",
    "pop = \"Sdekayi_p123_v4_25miss_popSdekayi\"\n",
    "# get the number of samples retained in the projection\n",
    "nsamps = proj_dict[pop] \n",
    "# get the full sequence length from the stats file\n",
    "in_stats = vcf_dir + pop.rsplit(\"_\", 1)[0] + \"_stats.txt\"\n",
    "seq_length = open(in_stats).readlines()[-1].split(\",\")[1].strip().split(\")\")[0]\n",
    "\n",
    "\n",
    "# get the folded sfs\n",
    "pop_sfs = analysis_dir + pop + \"/sfs/dadi/\" + pop.split(\"_pop\", 1)[1] + \".sfs\".format(pop=pop)\n",
    "pop_sfs = open(pop_sfs).readlines()[1].split()[1:int(nsamps/2)+1]\n",
    "\n",
    "# Write the blueprint file\n",
    "blueprint_file = analysis_dir + \"{pop}/{pop}.blueprint\".format(pop=pop)\n",
    "with open(blueprint_file, 'w') as outfile:\n",
    "    outfile.write(blueprint_template.format(pop=pop, nseqs=nsamps, seq_length=seq_length, sfs=\" \".join(pop_sfs)))\n",
    "\n",
    "# This call creates $blueprint_file.sh which gets run\n",
    "!java -cp $stairwayplot_es_dir Stairbuilder \"$blueprint_file\"\n",
    "stairwayplot2_script = blueprint_file + \".sh\"\n",
    "!bash \"$stairwayplot2_script\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdf634dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## do them all in a serial fashion, dumb but easy\n",
    "\n",
    "for pop in all_pops:\n",
    "    print(pop)\n",
    "    stairwayplot_es_dir = \"/Applications/stairway_plot_v2.1.1/stairway_plot_es\"\n",
    "    # Get the number of samples retained in the projection\n",
    "    nsamps = proj_dict[pop]\n",
    "    # get the full sequence length from the stats file\n",
    "    in_stats = vcf_dir + pop.rsplit(\"_\", 1)[0] + \"_stats.txt\"\n",
    "    seq_length = open(in_stats).readlines()[-1].split(\",\")[1].strip().split(\")\")[0]\n",
    "\n",
    "    # get the folded sfs\n",
    "    pop_sfs = analysis_dir + pop + \"/sfs/dadi/\" + pop.split(\"_pop\", 1)[1] + \".sfs\".format(pop=pop)\n",
    "    pop_sfs = open(pop_sfs).readlines()[1].split()[1:int(nsamps/2)+1]\n",
    "\n",
    "    # Write the blueprint file\n",
    "    blueprint_file = analysis_dir + \"{pop}/{pop}.blueprint\".format(pop=pop)\n",
    "    with open(blueprint_file, 'w') as outfile:\n",
    "        outfile.write(blueprint_template.format(pop=pop, nseqs=nsamps, seq_length=seq_length, sfs=\" \".join(pop_sfs)))\n",
    "    \n",
    "    # This call creates $blueprint_file.sh which gets run\n",
    "    _ = !java -cp $stairwayplot_es_dir Stairbuilder \"$blueprint_file\"\n",
    "    stairwayplot2_script = blueprint_file + \".sh\"\n",
    "    _ = !bash \"$stairwayplot2_script\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b0a04fe",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
